### PR TITLE
WIP: Proof of concept for ML in scilpy - LabelSeg

### PR DESCRIPTION
**Pull request by AntoineTheb**
# Quick description

Please include a summary of the changes and the related issue(s) or improvement(s).
Please also include relevant motivation and context. List any dependencies that are required for this change if needed.

...

## Type of change

Check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Provide data, screenshots, command line to test (if relevant)

...

# Checklist

- [ ] My code follows the style guidelines of this project (run [autopep8](https://pypi.org/project/autopep8/))
- [ ] I added relevant citations to scripts, modules and functions docstrings and descriptions
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I moved all functions from the script file (except the argparser and main) to scilpy modules
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: scripts/scil_labelseg.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+torch, have_torch, _ = optional_package('torch', trip_msg=IMPORT_ERROR_MSG)
+
+# TODO: Get bundle list from model
+DEFAULT_BUNDLES = ['AF_left', 'AF_right', 'ATR_left', 'ATR_right', 'CA', 'CC_1', 'CC_2', 'CC_3', 'CC_4', 'CC_5', 'CC_6', 'CC_7', 'CG_left', 'CG_right', 'CST_left', 'CST_right', 'FPT_left', 'FPT_right', 'FX_left', 'FX_right', 'ICP_left', 'ICP_right', 'IFO_left', 'IFO_right', 'ILF_left', 'ILF_right', 'MCP', 'MLF_left', 'MLF_right', 'OR_left', 'OR_right', 'POPT_left', 'POPT_right', 'SCP_left', 'SCP_right', 'SLF_III_left', 'SLF_III_right', 'SLF_II_left', 'SLF_II_right', 'SLF_I_left', 'SLF_I_right', 'STR_left', 'STR_right', 'ST_FO_left', 'ST_FO_right', 'ST_OCC_left', 'ST_OCC_right', 'ST_PAR_left', 'ST_PAR_right', 'ST_POSTC_left', 'ST_POSTC_right', 'ST_PREC_left', 'ST_PREC_right', 'ST_PREF_left', 'ST_PREF_right', 'ST_PREM_left', 'ST_PREM_right', 'T_OCC_left', 'T_OCC_right', 'T_PAR_left', 'T_PAR_right', 'T_POSTC_left', 'T_POSTC_right', 'T_PREC_left', 'T_PREC_right', 'T_PREF_left', 'T_PREF_right', 'T_PREM_left', 'T_PREM_right', 'UF_left', 'UF_right']  # noqa E501
+
+DEFAULT_CKPT = os.path.join('checkpoints', 'labelsegnet.ckpt')
```
Comment (levje) [inline scripts/scil_labelseg.py:46]: Maybe the default checkpoint could be the .scilpy/checkpoints/labelsegnet.ckpt. And maybe this path could be used by future projects too. Like .scilpy/checkpoints/tractoracle.ckpt.

File: scripts/scil_labelseg.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+        ).to(device)
+
+        wm_prompt = torch.tensor(
+            wm_data,
+            dtype=torch.float
+        ).to(device)
+
+        prompts = torch.eye(len(self.bundles), device=device)
+
+        with torch.no_grad():
```
Comment (levje) [inline scripts/scil_labelseg.py:161]: Probably don't need that with your decorator

File: scripts/scil_labelseg.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+DEFAULT_CKPT = os.path.join('checkpoints', 'labelsegnet.ckpt')
+
+
+def _build_arg_parser(parser):
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=RawTextHelpFormatter)
+
+    parser.add_argument('fodf', type=str,
+                        help='fODF input')
```
Comment (arnaudbore) [inline scripts/scil_labelseg.py:55]: in_fodf and explicit the order of sh here again.

File: scripts/scil_labelseg.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+from scilpy.io.utils import (
+    assert_inputs_exist, assert_outputs_exist, add_overwrite_arg)
+from scilpy.image.volume_operations import resample_volume
+
+from scilpy.ml.labelseg.predict import predict
+from scilpy.ml.labelseg.utils import get_data, get_model, download_weights
+from scilpy.ml.utils import get_device
+
+
+from dipy.utils.optpkg import optional_package
```
Comment (arnaudbore) [inline scripts/scil_labelseg.py:36]: check the order of your imports + add scilpy version.

File: scripts/scil_labelseg.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    parser.add_argument('fodf', type=str,
+                        help='fODF input')
+    parser.add_argument('wm', type=str,
+                        help='WM input')
+    parser.add_argument('out_prefix', type=str,
+                        help='Output destination and file prefix.')
+    parser.add_argument('--mask', type=str, default=None,
+                        help='Output destination and file prefix for '
+                             'binary mask output.')
+    parser.add_argument('--nb_labels', type=int, default=50)
```
Comment (arnaudbore) [inline scripts/scil_labelseg.py:63]: Needs a help

File: scripts/scil_labelseg.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+def _build_arg_parser(parser):
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=RawTextHelpFormatter)
+
+    parser.add_argument('fodf', type=str,
+                        help='fODF input')
+    parser.add_argument('wm', type=str,
+                        help='WM input')
```
Comment (arnaudbore) [inline scripts/scil_labelseg.py:57]: does it have to be a mask ?

File: scripts/scil_labelseg.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    parser.add_argument('out_prefix', type=str,
+                        help='Output destination and file prefix.')
+    parser.add_argument('--mask', type=str, default=None,
+                        help='Output destination and file prefix for '
+                             'binary mask output.')
```
Comment (arnaudbore) [inline scripts/scil_labelseg.py:62]: I don't get why we need a mask here.

File: scripts/scil_labelseg.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    parser.add_argument('--nb_labels', type=int, default=50)
+    parser.add_argument('--checkpoint', type=str,
+                        default=DEFAULT_CKPT,
+                        help='Checkpoint (.ckpt) containing hyperparameters '
+                             'and weights of model. Default is '
+                             '[%(default)s].')
```
Comment (arnaudbore) [inline scripts/scil_labelseg.py:68]: If not set it will be downloaded

File: scripts/scil_labelseg.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python
+
+"""
+LabelSeg: automatic tract labelling without tractography.
```
Comment (arnaudbore) [inline scripts/scil_labelseg.py:4]: we should come up with a better name _maybe_ scil_fodf_labelseg.py

**Review by arnaudbore — COMMENTED** [review]
I love it, I started the review as a real one before realizing it's still a WIP and there are a lot of TODOs. I'm going to give it a try but overall it's very light and it's a confirmation that ml dev should be usable through scilpy :) Great Job ! Thank you !

File: scilpy/ml/labelseg/utils.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+def get_data(fodf, wm, img_size, n_coefs):
+
+    # Instanciate an empty numpy array to store the fodf data
+    fodf_data = np.zeros((n_coefs, *fodf.shape[:3]))
+
+    # Select the first n_coefs coefficients from the fodf data
+    # and put it in the first dimension
+    # This should be robust to higher or lower numbers of coefficients
+    # than the model was trained on
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scilpy/ml/labelseg/utils.py:24]: fODF of order > 6 (ie 8) could be truncated and still work. I *was* hoping that by padding fODFs of order < 6 (ie 4) I could get it to work reasonably well, but I haven't tested it. Prob. safer to just prevent the user from inputting fODF of order < 6.

File: scilpy/ml/labelseg/encodings.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+torch, have_torch, _ = optional_package('torch', trip_msg=IMPORT_ERROR_MSG)
+
+
+# From https://github.com/tatp22/multidim-positional-encoding
+def get_emb(sin_inp):
+    """
+    Gets a base embedding for one dimension with sin and cos intertwined
+    """
+    emb = torch.stack((sin_inp.sin(), sin_inp.cos()), dim=-1)
+    return torch.flatten(emb, -2, -1)
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scilpy/ml/labelseg/encodings.py:17]: Sure, but let's cross that bridge when we get to it.

File: scripts/scil_labelseg.py (hunk 9 above)
Comment (AntoineTheb) [reply scripts/scil_labelseg.py:4]: Still open to suggestions ! I've stuck with labelseg so far only because I haven't gotten a better alternative.

File: scripts/scil_labelseg.py (hunk 7 above)
Comment (AntoineTheb) [reply scripts/scil_labelseg.py:62]: The method takes in as input a whole-brain WM mask and can output a bundle-wise binary mask. I'll make this clearer in the doc.

Comment (AntoineTheb) [reply scripts/scil_labelseg.py:62]: Actually I could also just remove the option (I've kept it since we get it "for free" as the model outputs one) as it can be obtained by binarizing the label map.

**Comment by EmmaRenauld** [comment]
Was this discussed with Maxime? We once had a discussion about the idea of including ML in scilpy, if it was a good idea or not, since generally, it will not be reviewed or understood by the same developpers as the rest of scilpy. That's a bit why, two years ago, we decided to create dwiml instead of including it in scilpy. Should it go in dwiml?

File: scilpy/ml/labelseg/labelsegnet.py [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+        self.conv2 = torch.nn.Conv3d(
+            in_chans, in_chans * ratio, kernel_size=1, stride=1)
+        self.gelu1 = torch.nn.GELU()
+
+        self.conv3 = torch.nn.Conv3d(
+            in_chans * ratio, in_chans, kernel_size=1, stride=1)
+
+        # Use Xavier initialisation for weights
+        for m in self.modules():
+            if isinstance(m, torch.nn.Conv3d) or isinstance(m, torch.nn.ConvTranspose3d):
```
Comment (levje) [inline scilpy/ml/labelseg/labelsegnet.py:177]: Probably doesn't matter, but there's no torch.nn.ConvTranspose3d in your ConvNextBlock and DownsampleBlock. Maybe a common function to init the weights of all the blocks could be cool. If you don't agree, this looks clean either way.

File: scilpy/ml/labelseg/labelsegnet.py [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+            self.pe_layer = PositionalEncodingPermute3D(out_chans)
+            self._prompt_func = self._prompt_attn
+        else:
+            self.prompt_conv = ConvNextBlock(out_chans, ratio)
+            self._prompt_func = self._prompt_add
+
+        self.ds_head = Head(out_chans)
+
+    def _decode(self, z, encoder_feature):
+        """ TODO """
```
Comment (levje) [inline scilpy/ml/labelseg/labelsegnet.py:339]: TODO, same for the other functions below

File: scilpy/ml/labelseg/labelsegnet.py [hunk 14]
```diff
@@ -0,0 +1,586 @@
...
+        super().__init__()
+
+        self.prompt_strategy = prompt_strategy
+
+        self.upsample = UpsampleNextBlock(
+            in_chans, out_chans, ratio)
+        self.conv1 = ConvNextBlock(out_chans, ratio)
+        self.conv2 = ConvNextBlock(out_chans, ratio)
+
+        # TODO: Split the prompting strategy into a separate class
```
Comment (levje) [inline scilpy/ml/labelseg/labelsegnet.py:323]: Can this todo be removed?

File: scilpy/ml/labelseg/labelsegnet.py [hunk 15]
```diff
@@ -0,0 +1,586 @@
...
+
+
+class Head(torch.nn.Module):
+
+    def __init__(self, in_chans: int):
+        super().__init__()
+
+        self.conv = torch.nn.Conv3d(
+            in_chans, 2, kernel_size=1, stride=1)
+        # self.act = torch.nn.Sigmoid()
```
Comment (levje) [inline scilpy/ml/labelseg/labelsegnet.py:442]: unused sigmoid, also a line in the forward

File: scilpy/ml/labelseg/labelsegnet.py [hunk 16]
```diff
@@ -0,0 +1,586 @@
...
+        self.stem = Stem(in_chans, embed_dim)
+        self.mask_stem = Stem(1, embed_dim)
+
+        self.encoder = UNextEncoder(self.channels)
+        self.mask_encoder = UNextEncoder(self.channels)
+
+        self.bottleneck = ConvNextBlock(bottleneck_dim, ratio=4)
+        self.decoder = LabelSegNetDecoder(prompt_strategy,
+                                          channels=self.channels[::-1])
+        # self.head = Head(embed_dim)
```
Comment (levje) [inline scilpy/ml/labelseg/labelsegnet.py:477]: unused

File: scilpy/ml/labelseg/predict.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+    mask : np.ndarray
+        Predicted mask for the bundle.
+    bundle_name : str
+        Name of the bundle.
+    """
+
+    # Binarize the mask
+    bundle_mask = (mask > 0.5)
+
+    # TODO: investigate blobs. TractSeg performs some post-processing
```
Comment (levje) [inline scilpy/ml/labelseg/predict.py:30]: Most of this is commented, could we remove it?

File: scilpy/ml/utils.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+                   " it first. See the official website for more info: " + \
+                   "https://pytorch.org/get-started/locally/" # noqa
+torch, have_torch, _ = optional_package('torch', trip_msg=IMPORT_ERROR_MSG)
+
+
+def get_device():
+    if torch.cuda.is_available():
+        return torch.device("cuda")
+    elif torch.backends.mps.is_available():
+        return torch.device("mps")
```
Comment (levje) [inline scilpy/ml/utils.py:14]: As much as I like this "mps" backend, I'm not sure it should be included in scilpy for now. There's still some stuff that is not implemented on that backend and when you encounter that, it raises a weird unreadable error. For example, just having a BatchNorm3d is problematic.

Or, maybe we can add a warning that states that "Some functionalities might not be supported by the "mps" backend" whenever this gets imported for the first time. This might be more confusing than helpful though.

Comment (AntoineTheb) [reply scilpy/ml/utils.py]: Easier to just remove it.

File: scilpy/ml/labelseg/predict.py (hunk 17 above)
Comment (AntoineTheb) [reply scilpy/ml/labelseg/predict.py:30]: I modified it to leave a note to my future-self.

File: scilpy/ml/labelseg/labelsegnet.py (hunk 12 above)
Comment (AntoineTheb) [reply scilpy/ml/labelseg/labelsegnet.py]: Done

**Comment by AntoineTheb** [comment]
@EmmaRenauld
> Was this discussed with Maxime? We once had a discussion about the idea of including ML in scilpy, if it was a good idea or not, since generally, it will not be reviewed or understood by the same developpers as the rest of scilpy. That's a bit why, two years ago, we decided to create dwiml instead of including it in scilpy. Should it go in dwiml?

We've (@arnaudbore, @mdesco, @levje andI) indeed discussed it extensively recently offline . One of the arguments for the inclusion was that including stuff in scilpy faciliates the integration of ML tools in flows, for example. This PR is a proof of concept for integrating "inference stuff" only, keeping the code as light as possible.

There is indeed an issue with dependencies, but one good point was that we already have some scripts that are dependent on OpenCL. Right now OpenCL is an optional dependency and so could Pytorch.

**Review by levje — APPROVED** [review]

**Comment by AntoineTheb** [comment]
@arnaudbore ready to be reviewed again ! Do you have any clues wrt the code coverage issues ?

File: scripts/scil_fodf_bundleparc.py [hunk 19]
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python
+
+"""
+BundleParc: automatic tract labelling without tractography.
+
+This method takes as input fODF maps of order 6 (or above) and a whole-brain WM mask and outputs 71 bundle label maps. These maps can then be used to perform tractometry/tract profiling/radiomics. The bundle definitions follow TractSeg's minus the whole CC.
+
+Example usage:
+    $ scil_bundleparc.py fodf.nii.gz wm_mask.nii.gz --out_prefix sub-001__
```
Comment (arnaudbore) [inline scripts/scil_fodf_bundleparc.py:9]:
```suggestion
    $ scil_fodf_bundleparc.py fodf.nii.gz wm_mask.nii.gz --out_prefix sub-001__
```

File: scripts/scil_fodf_bundleparc.py [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+    resampled_wm = resample_volume(wm_in, ref_img=None,
+                                   volume_shape=[img_size],
+                                   iso_min=False,
+                                   voxel_res=None,
+                                   interp='nn',
+                                   enforce_dimensions=False)
+    # Hack so that get_data_as_mask works
+    resampled_wm.set_filename(wm_in.get_filename())
+
+    # shape = fodf_in.get_fdata().shape[:3]
```
Comment (levje) [inline scripts/scil_fodf_bundleparc.py:141]: Maybe remove this commented line

**Review by levje — APPROVED** [review]
lgtm

**Review by arnaudbore — APPROVED** [review]
